api/channel.go

```diff
@@ -360,43 +360,62 @@ func JoinChannel(c *Context, channelId string, role string) {
 		channel := cresult.Data.(*model.Channel)
 		user := uresult.Data.(*model.User)
 
-		if !c.HasPermissionsToTeam(channel.TeamId, "joinChannel") {
-			return
-		}
-
-		if channel.DeleteAt > 0 {
-			c.Err = model.NewAppError("joinChannel", "The channel has been archived or deleted", "")
-			c.Err.StatusCode = http.StatusBadRequest
+		if !c.HasPermissionsToTeam(channel.TeamId, "join") {
 			return
 		}
 
 		if channel.Type == model.CHANNEL_OPEN {
-			cm := &model.ChannelMember{ChannelId: channel.Id, UserId: c.Session.UserId,
-				Roles: role, NotifyProps: model.GetDefaultChannelNotifyProps()}
-
-			if cmresult := <-Srv.Store.Channel().SaveMember(cm); cmresult.Err != nil {
-				c.Err = cmresult.Err
-				return
-			}
-
-			post := &model.Post{ChannelId: channel.Id, Message: fmt.Sprintf(
-				`User %v has joined this channel.`,
-				user.Username), Type: model.POST_JOIN_LEAVE}
-			if _, err := CreatePost(c, post, false); err != nil {
-				l4g.Error("Failed to post join message %v", err)
-				c.Err = model.NewAppError("joinChannel", "Failed to send join request", "")
+			if _, err := AddUserToChannel(user, channel); err != nil {
+				c.Err = err
 				return
 			}
-
-			UpdateChannelAccessCacheAndForget(c.Session.TeamId, c.Session.UserId, channel.Id)
+			PostUserAddRemoveMessageAndForget(c, channel.Id, fmt.Sprintf(`User %v has joined this channel.`, user.Username))
 		} else {
-			c.Err = model.NewAppError("joinChannel", "You do not have the appropriate permissions", "")
+			c.Err = model.NewAppError("join", "You do not have the appropriate permissions", "")
 			c.Err.StatusCode = http.StatusForbidden
 			return
 		}
 	}
 }
 
+func PostUserAddRemoveMessageAndForget(c *Context, channelId string, message string) {
+	go func() {
+		post := &model.Post{
+			ChannelId: channelId,
+			Message:   message,
+			Type:      model.POST_JOIN_LEAVE,
+		}
+		if _, err := CreatePost(c, post, false); err != nil {
+			l4g.Error("Failed to post join/leave message %v", err)
+		}
+	}()
+}
+
+func AddUserToChannel(user *model.User, channel *model.Channel) (*model.ChannelMember, *model.AppError) {
```

Comment:
nvm, I see now JoinChannel was renamed to join.  I didn't see it in the git merge.